srtcore/sync.cpp

```diff
@@ -0,0 +1,168 @@
+    uint32_t lval, hval;
+    asm("rdtsc" : "=a"(lval), "=d"(hval));
+    x = hval;
+    x = (x << 32) | lval;
+#elif defined(TIMING_USE_QPC)
+    // This function should not fail, because we checked the QPC
+    // when calling to QueryPerformanceFrequency. If it failed,
+    // the m_bUseMicroSecond was set to true.
+    QueryPerformanceCounter((LARGE_INTEGER*)&x);
+#elif defined(TIMING_USE_MACH_ABS_TIME)
+    x = mach_absolute_time();
+#else
+    // use system call to read time clock for other archs
+    timeval t;
+    gettimeofday(&t, 0);
+    x = t.tv_sec * uint64_t(1000000) + t.tv_usec;
+#endif
+}
+
+int64_t get_cpu_frequency()
+{
+    int64_t frequency = 1; // 1 tick per microsecond.
+
+#if defined(TIMING_USE_QPC)
+
+    LARGE_INTEGER ccf; // in counts per second
+    if (QueryPerformanceFrequency(&ccf))
+        frequency = ccf.QuadPart / 1000000; // counts per microsecond
+
+#elif defined(TIMING_USE_MACH_ABS_TIME)
+
+    mach_timebase_info_data_t info;
+    mach_timebase_info(&info);
+    frequency = info.denom * int64_t(1000) / info.numer;
+
+#elif defined(IA32) || defined(IA64) || defined(AMD64)
+    uint64_t t1, t2;
+
+    rdtsc(t1);
+    timespec ts;
+    ts.tv_sec  = 0;
+    ts.tv_nsec = 100000000;
+    nanosleep(&ts, NULL);
+    rdtsc(t2);
+
+    // CPU clocks per microsecond
+    frequency = int64_t(t2 - t1) / 100000;
+#endif
+
+    return frequency;
+}
+
+static const int64_t s_cpu_frequency = get_cpu_frequency();
+
+} // namespace sync
+} // namespace srt
+
+template <>
+uint64_t srt::sync::TimePoint<srt::sync::steady_clock>::us_since_epoch() const
+{
+    return m_timestamp / s_cpu_frequency;
+}
+
+template <>
+srt::sync::Duration<srt::sync::steady_clock> srt::sync::TimePoint<srt::sync::steady_clock>::time_since_epoch() const
+{
+    return srt::sync::Duration<srt::sync::steady_clock>(m_timestamp);
+}
+
+srt::sync::TimePoint<srt::sync::steady_clock> srt::sync::steady_clock::now()
+{
+    uint64_t x = 0;
+    rdtsc(x);
+    return TimePoint<steady_clock>(x);
+}
+
+long long srt::sync::count_microseconds(const steady_clock::duration& t) { return t.count() / s_cpu_frequency; }
+
+long long srt::sync::count_milliseconds(const steady_clock::duration& t) { return t.count() / s_cpu_frequency / 1000; }
+
+long long srt::sync::count_seconds(const steady_clock::duration& t) { return t.count() / s_cpu_frequency / 1000000; }
+
+srt::sync::steady_clock::duration srt::sync::microseconds_from(long t_us)
+{
+    return steady_clock::duration(t_us * s_cpu_frequency);
+}
+
+srt::sync::steady_clock::duration srt::sync::milliseconds_from(long t_ms)
+{
+    return steady_clock::duration((1000 * t_ms) * s_cpu_frequency);
+}
+
+srt::sync::steady_clock::duration srt::sync::seconds_from(long t_s)
+{
+    return steady_clock::duration((1000000 * t_s) * s_cpu_frequency);
+}
+
+std::string srt::sync::FormatTime(const steady_clock::time_point& timestamp)
+{
+    const time_t                   now_s         = ::time(NULL); // get current time in seconds
```

Comment:
Should be using clock_gettime(CLOCK_MONOTONIC,..) if ENABLE_MONOTONIC_CLOCK is enabled: ':time()' may jump but 'steady_clock::now()' should not. The result will be misleading timestamp formating.

Reply:
This `::time` is taken in order to convert the value of the steady clock into the system clock - so actually it would be monotonic clock completely wrong here. Actually the value from monotonic clock is taken in the line below. And both are taken for a reason: this function must convert steady clock into system clock.

BTW. @maxsharabayko : I have doubts as to whether this is the most optimal method to convert the steady clock into the system clock. Could you check this extra please?

Reply:
@ethouris
> I have doubts as to whether this is the most optimal method to convert the steady clock into the system clock. Could you check this extra please?

What is your suggestion?

Reply:
Usually if a monotonic clock and system clock differ (and still can be counted in the same units), it means that they have simply different bases (points in time when the time value is zero). Maybe there is a method to confront their bases, which will be less effort consuming and more reliable than taking the current time in both clocks.

Reply:
Why do we care about converting SRT time (hopefully monotonic) to actual system time in the logs? SRT time is what we care about for troubleshooting and tracing, converting to a different base makes it confusing and potentially broken: we will not be able to trust our own traces because we can't guarantee that the system time has not jumped.

Reply:
How do you propose to display the time in the logs then?

Reply:
What we display should use the same time base as what we use in the code to make timing decisions. Otherwise I don't see any value in tracing the time. The logger can display system time in the beginning of log entries.

Reply:
Good point!
- [x] Renamed the previous implementation of FormatTime to FormatTimeSys.
Example output: `12:19:52.370140 [SYS]`

- [x] Added new FormatTime implementation, that converts steady clock timestamp to a human-readable representation (losing precision below microseconds)
Example output: `2D 02:20:33.365191 [STD]`